keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,386 @@
+
+```golang
+type NodeStatus struct {
+        …
+        ComputeResources []ComputeResource
+        …
+}
+type ComputeResource struct {
+        // unique and deterministically generated. “nodeName-resourceName-propertyHash” naming convention,
+        // where propertyHash is generated by calculating a hash over all resource properties
+        Name string
+        // raw resource name. E.g.: nvidia.com/nvidia-gpu
+        ResourceName string
+        // resource metadata received from device plugin.
+        // e.g., gpuType: k80, zone: us-west1-b
+        Properties map[string]string
+        // list of deviceIds received from device plugin.
+        // e.g., ["nvida0", "nvidia1"]
+        Devices []string
+        // similar to the above but only contains allocatable devices.
+        AllocatableDevices []string
+}
+```
+The ComputeResource name needs to be unique and deterministically generated. We propose to use “nodeName-resourceName-propertyHash” naming convention for node level resources, where propertyHash is generated by calculating a hash over all resource properties. The properties of a physical resource may change, e.g., through a driver or node upgrade. When this happens, Kubelet will need to create a new ComputeResource object and delete the old one. However, some pods may already be scheduled on the node and allocated with the old ComputeResource, and it is hard for the scheduler to do its bookkeeping when such pods are still running on the node while the associated ComputeResource is removed. We have a few options to handle this situation.
+- First, we may document that to change or remove a device resource, the node has to be drained. This requirement however may complicate device plugin upgrade process.
+- Another option is that Kubelet can evict the pods that are allocated with an unexisting ComputeResource. Although simple, this approach may disturb long-running workloads during device plugin upgrade.
+- To support a less disruptive model, upon resource property change, Kubelet can still export capacity at old ComputeResource name for the devices used by active pods, and exports capacity at new matching ComputeResource name for devices not in use. Only when those pods finish running, that particular node finishes its transition. This approach avoids resource multiple counting and simplifies the scheduler resource accounting. One potential downside is that the transition may take quite long process if there are long running pods using the resource on the nodes. In that case, cluster admins can still drain the node at convenient time to speed up the transition. Note that this approach does add certain code complexity on Kubelet DeviceManager component.
+
+Possible fields we may consider to add later include:
+- `DeviceUnits resource.Quantity`. This field can be used to support fractional
+  resource or infinite resource. In a more advanced use case, a device plugin may
```

Comment:
Infinite resources might come handy in the case when a "default" non-countable resource should be assigned to a container. This could be used for example to make a DP set a node-specific environment variable to a container.